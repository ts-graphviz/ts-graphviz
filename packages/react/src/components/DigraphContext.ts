import React from 'react';
import { Digraph } from 'ts-graphviz';
export interface IDigraphContext {
  g: Digraph;
}

export const DigraphContext = React.createContext<IDigraphContext>({
  g: new Digraph(),
});
