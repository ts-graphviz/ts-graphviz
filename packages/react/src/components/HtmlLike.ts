/* eslint-disable no-undef */
import { AttributesValue } from 'ts-graphviz';

export type TableProps = {
  ALIGN?: 'CENTER' | 'LEFT' | 'RIGHT'; // "CENTER|LEFT|RIGHT"
  BGCOLOR?: AttributesValue; // "color"
  BORDER?: AttributesValue; // "value"
  CELLBORDER?: AttributesValue; // "value"
  CELLPADDING?: AttributesValue; // "value"
  CELLSPACING?: AttributesValue; // "value"
  COLOR?: AttributesValue; // "color"
  COLUMNS?: AttributesValue; // "value"
  FIXEDSIZE?: true; // "FALSE|TRUE"
  GRADIENTANGLE?: AttributesValue; // "value"
  HEIGHT?: AttributesValue; // "value"
  HREF?: AttributesValue; // "value"
  ID?: AttributesValue; // "value"
  PORT?: AttributesValue; // "portName"
  ROWS?: AttributesValue; // "value"
  SIDES?: AttributesValue; // "value"
  STYLE?: AttributesValue; // "value"
  TARGET?: AttributesValue; // "value"
  TITLE?: AttributesValue; // "value"
  TOOLTIP?: AttributesValue; // "value"
  VALIGN?: 'MIDDLE' | 'BOTTOM' | 'TOP'; // "MIDDLE|BOTTOM|TOP"
  WIDTH?: AttributesValue; // "value"
};

type NoAttributes = {};

export type TrProps = NoAttributes;

export type TdProps = {
  ALIGN?: 'CENTER' | 'LEFT' | 'RIGHT' | 'TEXT'; // "CENTER|LEFT|RIGHT|TEXT"
  BALIGN?: 'CENTER' | 'LEFT' | 'RIGHT'; // "CENTER|LEFT|RIGHT"
  BGCOLOR?: AttributesValue; // "color"
  BORDER?: AttributesValue; // "value"
  CELLPADDING?: AttributesValue; // "value"
  CELLSPACING?: AttributesValue; // "value"
  COLOR?: AttributesValue; // "color"
  COLSPAN?: AttributesValue; // "value"
  FIXEDSIZE?: boolean; // "FALSE|TRUE"
  GRADIENTANGLE?: AttributesValue; // "value"
  HEIGHT?: AttributesValue; // "value"
  HREF?: AttributesValue; // "value"
  ID?: AttributesValue; // "value"
  PORT?: AttributesValue; // "portName"
  ROWSPAN?: AttributesValue; // "value"
  SIDES?: AttributesValue; // "value"
  STYLE?: AttributesValue; // "value"
  TARGET?: AttributesValue; // "value"
  TITLE?: AttributesValue; // "value"
  TOOLTIP?: AttributesValue; // "value"
  VALIGN?: 'MIDDLE' | 'BOTTOM' | 'TOP'; // "MIDDLE|BOTTOM|TOP"
  WIDTH?: AttributesValue; // "value"
};

export type FontProps = {
  color?: AttributesValue; // "color"
  face?: AttributesValue; // "fontname"
  point?: AttributesValue; // SIZE="value"
};

export type BrProps = {
  ALIGN?: 'CENTER' | 'LEFT' | 'RIGHT'; // "CENTER|LEFT|RIGHT"
};

export type ImgProps = {
  SCALE?: boolean | 'WIDTH' | 'HEIGHT' | 'BOTH'; // "FALSE|TRUE|WIDTH|HEIGHT|BOTH"
  SRC?: AttributesValue; // "value"
};

export type IProps = NoAttributes;

export type BProps = NoAttributes;

export type UProps = NoAttributes;

export type OProps = NoAttributes;
export type SubProps = NoAttributes;

export type SupProps = NoAttributes;
export type SProps = NoAttributes;
export type HrProps = NoAttributes;
export type VrProps = NoAttributes;

export enum DOT {
  PORT = 'dot-port',
  TABLE = 'dot-table',
  TR = 'dot-tr',
  TD = 'dot-td',
  FONT = 'dot-font',
  BR = 'dot-br',
  IMG = 'dot-img',
  I = 'dot-i',
  B = 'dot-b',
  U = 'dot-u',
  O = 'dot-o',
  SUB = 'dot-sub',
  SUP = 'dot-sup',
  S = 'dot-s',
  HR = 'dot-hr',
  VR = 'dot-vr',
}

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    interface IntrinsicElements {
      [DOT.PORT]: { children: string };
      [DOT.TABLE]: React.PropsWithChildren<TableProps>;
      [DOT.TR]: React.PropsWithChildren<TrProps>;
      [DOT.TD]: React.PropsWithChildren<TdProps>;
      [DOT.FONT]: React.PropsWithChildren<FontProps>;
      [DOT.BR]: BrProps;
      [DOT.IMG]: ImgProps;
      [DOT.I]: React.PropsWithChildren<IProps>;
      [DOT.B]: React.PropsWithChildren<BProps>;
      [DOT.U]: React.PropsWithChildren<UProps>;
      [DOT.O]: React.PropsWithChildren<OProps>;
      [DOT.SUB]: React.PropsWithChildren<SubProps>;
      [DOT.SUP]: React.PropsWithChildren<SupProps>;
      [DOT.S]: React.PropsWithChildren<SProps>;
      [DOT.HR]: HrProps;
      [DOT.VR]: VrProps;
    }
  }
}
