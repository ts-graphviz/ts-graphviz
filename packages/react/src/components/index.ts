export * from './Digraph';
export * from './Edge';
export * from './Graph';
export * from './Node';
export * from './Subgraph';
