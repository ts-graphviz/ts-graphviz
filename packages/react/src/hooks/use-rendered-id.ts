import { ReactElement, useMemo } from 'react';
import { renderId } from '../utils/render-id';

export function useRenderedID(id?: ReactElement | string): string | undefined {
  return useMemo(() => renderId(id), [id]);
}
