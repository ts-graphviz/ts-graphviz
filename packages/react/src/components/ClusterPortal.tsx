import React, { FC, useContext, useMemo } from 'react';
import PropTypes from 'prop-types';
import { ICluster } from 'ts-graphviz';
import { Cluster } from './contexts/Cluster';
import { ClusterMap } from './contexts/ClusterMap';
import { useRootCluster } from '../hooks/use-root-cluster';

type Props = {
  name?: string;
};

export const ClusterPortal: FC<Props> = ({ children, name }) => {
  const root = useRootCluster();
  const map = useContext(ClusterMap);
  const cluster = useMemo(() => {
    if (name && map.has(name)) {
      return map.get(name) as ICluster;
    }
    return root;
  }, [root, map, name]);
  return <Cluster.Provider value={cluster}>{children}</Cluster.Provider>;
};

ClusterPortal.displayName = 'ClusterPortal';
ClusterPortal.defaultProps = {
  name: undefined,
};

ClusterPortal.propTypes = {
  name: PropTypes.string,
};
