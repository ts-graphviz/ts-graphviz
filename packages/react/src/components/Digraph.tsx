import React, { FC, ReactElement } from 'react';
import PropTypes from 'prop-types';
import { RootCluster, NoRootCluster } from './contexts/RootCluster';
import { Cluster } from './contexts/Cluster';
import { useDigraph, DigraphProps } from '../hooks/use-digraph';
import { useRenderedID } from '../hooks/use-rendered-id';
import { useRootCluster } from '../hooks/use-root-cluster';
import { DuplicatedRootClusterErrorMessage } from '../utils/errors';

type Props = Omit<DigraphProps, 'label'> & {
  label?: ReactElement | string;
};

export const Digraph: FC<Props> = ({ children, label, ...props }) => {
  const root = useRootCluster();
  if (root !== NoRootCluster) {
    throw Error(DuplicatedRootClusterErrorMessage);
  }
  const renderedLabel = useRenderedID(label);
  if (renderedLabel !== undefined) Object.assign(props, { label: renderedLabel });
  const digraph = useDigraph(props);
  return (
    <RootCluster.Provider value={digraph}>
      <Cluster.Provider value={digraph}>{children}</Cluster.Provider>
    </RootCluster.Provider>
  );
};

Digraph.displayName = 'Digraph';

Digraph.defaultProps = {
  id: undefined,
  comment: undefined,
  label: undefined,
};

Digraph.propTypes = {
  id: PropTypes.string,
  comment: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
};
