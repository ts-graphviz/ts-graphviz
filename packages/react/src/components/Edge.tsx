import React, { FC, useEffect, useMemo, ReactElement, isValidElement } from 'react';
import PropTypes from 'prop-types';
import { EdgeAttributes, EdgeTargetLike } from 'ts-graphviz';
import { renderToStaticMarkup } from 'react-dom/server';
import { useCluster } from '../hooks/useCluster';

type ReactEdgeAttributes = Omit<EdgeAttributes, 'label'> & {
  label?: ReactElement | string;
};

function renderId(id?: ReactElement | string): string | undefined {
  if (typeof id === 'string') {
    return id;
  }
  if (isValidElement(id)) {
    return renderToStaticMarkup(id);
  }
  return undefined;
}

type Props = {
  targets: EdgeTargetLike[];
  comment?: string;
} & ReactEdgeAttributes;

export const Edge: FC<Props> = ({ children, targets, comment, ...attributes }) => {
  const cluster = useCluster();
  const edge = useMemo(() => cluster.createEdge(...targets), [cluster, targets]);
  useEffect(() => {
    const { label, ...rest } = attributes;
    edge.attributes.clear();
    edge.attributes.apply({ label: renderId(label), ...rest });
  }, [edge, attributes]);

  useEffect(() => {
    edge.comment = comment;
  }, [edge, comment]);

  useEffect(() => {
    return (): void => {
      cluster.removeEdge(edge);
    };
  }, [cluster, edge]);
  return <>{children}</>;
};

Edge.displayName = 'Edge';

Edge.propTypes = {
  targets: PropTypes.array.isRequired,
  comment: PropTypes.string,
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
};

Edge.defaultProps = {
  comment: undefined,
  label: undefined,
};
