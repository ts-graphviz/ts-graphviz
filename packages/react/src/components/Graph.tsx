import React, { FC, useMemo } from 'react';
import gv from 'ts-graphviz';
import { ClusterContext } from '../contexts/ClusterContext';
import { useGraphvizContext } from '../hooks/useContext';

export const Graph: FC = ({ children }) => {
  const context = useGraphvizContext();
  const g = useMemo(() => new gv.Graph(context), [context]);
  return <ClusterContext.Provider value={g}>{children}</ClusterContext.Provider>;
};
