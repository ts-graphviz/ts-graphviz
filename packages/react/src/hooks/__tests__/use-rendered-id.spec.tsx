import React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { context } from './utils/wrapper';
import { useRenderedID } from '../use-rendered-id';
import { DOT } from '../../components/HtmlLike';

describe('useRenderedID', () => {
  test.each([
    'hoge',
    '<<b>bold</b>>',
    <DOT.B>bold</DOT.B>,
    <DOT.TABLE BORDER="0" CELLBORDER="1" CELLSPACING="0">
      <DOT.TR>
        <DOT.TD>left</DOT.TD>
        <DOT.TD PORT="m">middle</DOT.TD>
        <DOT.TD PORT="r">right</DOT.TD>
      </DOT.TR>
    </DOT.TABLE>,
  ])('', (id) => {
    const { result } = renderHook(() => useRenderedID(id), {
      wrapper: context(),
    });
    expect(result.current).toMatchSnapshot();
  });
});
