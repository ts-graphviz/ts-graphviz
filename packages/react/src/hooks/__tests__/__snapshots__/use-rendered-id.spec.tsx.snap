// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRenderedID  1`] = `"hoge"`;

exports[`useRenderedID  2`] = `"<<b>bold</b>>"`;

exports[`useRenderedID  3`] = `"<<b>bold</b>>"`;

exports[`useRenderedID  4`] = `"<<table BORDER=\\"0\\" CELLBORDER=\\"1\\" CELLSPACING=\\"0\\"><tr><td>left</td><td PORT=\\"m\\">middle</td><td PORT=\\"r\\">right</td></tr></table>>"`;
